### ISSUE TITLE
Why PBS can not count the memory usage correctly when work with MPI

### ISSUE THREAD

**Issue by louisXW** [PBSPro/pbspro#1052]
Here is a simple experiment I tried:

Given a task called "sim.exe" which doing a model simulation, I then use MPI to launch x "sim.exe" simultaneously on one node (shared memory system). I have tried four different runs with x to be different value (e.g., 1, 4, 8, 16). Then I check the memory usage through the PBS report "memory used" and "Vmem used". I observed that for these different runs the "memory used" and "Vmem used" keep the same not changed with "mem" = 8,432 KB and "vmem" = 489,716 KB.

However, if I do not use MPI but other parallelization techniques "threading" or "multiprocessing" in python, and launch x "sim.exe" simultaneously, the "mem" and "vmem" usage count returned by PBS is increasing linearly with the value of x, which is more as what I would expect.

So why "mem" and "vmem"  keeps the same even though there are more workload is doing? what does the "mem used" and "vmem used" return by PBS report means when MPI is used?

Can anyone provide some answers or suggestions on this?

The original post is here: [How to understand PBS output “mem” and “vmem” keep the same when the task is x-fold increased with mpirun -np x task](https://unix.stackexchange.com/questions/508288/how-to-understand-pbs-output-mem-and-vmem-keep-the-same-when-the-task-is-x-f)

**Comment by agrawalravi90** [comment]
Which implementation of MPI are you using? PBS can only track resource usage for MPI processes when using a flavour of MPI that it is integrated with. Please refer to section 10.1 ("Integration with MPI") of the PBS Admin guide: https://www.pbsworks.com/pdfs/PBSAdminGuide18.2.3.pdf

**Comment by louisXW** [comment]
The MPI I used is "Intel (R) MPI Library for Linux* OS, Version 2019" under the intel composer xe 2019 version. Can this be integrated with PBS? If not what the "mem" and "vmem" returned by PBS stand for?

Thanks!

**Comment by agrawalravi90** [comment]
Intel MPI should be integrated with PBS. Just reading this out of the Admin guide (section 10.5), can you check if the following variables have been set inside the $PBS_HOME/pbs_environment file:
I_MPI_HYDRA_BOOTSTRAP=rsh
I_MPI_HYDRA_BOOTSTRAP_EXEC=$PBS_EXEC/bin/pbs_tmrsh

Please replace $PBS_HOME and $PBS_EXEC with paths to your PBS_HOME and PBS_EXEC directories.

**Comment by louisXW** [comment]
I searched where the PBS was installed on the host. however, I could not find the "pbs_tmrsh"  Also where can I find the "pbs_envrionment file"?

**Comment by agrawalravi90** [comment]
Can you please cat /etc/pbs.conf and paste the output here? that will tell us where your PBS_HOME and PBS_EXEC are located

**Comment by louisXW** [comment]
[e0022672@nus02 /]$ cat /etc/pbs.conf
PBS_EXEC=/opt/pbs
PBS_HOME=/var/spool/pbs
PBS_START_SERVER=0
PBS_START_MOM=0
PBS_START_SCHED=0
PBS_START_COMM=0
PBS_SERVER=wlm01
PBS_PRIMARY=wlm01
PBS_SECONDARY=wlm02
PBS_CORE_LIMIT=unlimited
PBS_SCP=/usr/bin/scp
PBS_SENDMAIL=/usr/bin/sendmail_wrap
PBS_LOCALLOG=1
PBS_SYSLOG=5
PBS_SYSLOGSEVR=7

Thanks.

**Comment by agrawalravi90** [comment]
Thanks. Your pbs_environment file should be located at /var/spool/pbs/pbs_environment, and your pbs_tmrsh should be located at /opt/pbs/bin/pbs_tmrsh. Can you please check if I_MPI_HYDRA_BOOTSTRAP and I_MPI_HYDRA_BOOTSTRAP_EXEC are set correctly in your pbs_environment file?

**Comment by louisXW** [comment]
The cat of pbs_environment file gives:
```
[e0022672@nus02 PBS]$ cat pbs_environment
PATH=/bin:/usr/bin
TZ=Asia/Singapore
```
There is no pbs_tmrsh under the /opt/pbs/bin/

However, I observed there are "pbs_mpirun" what's that?

**Comment by agrawalravi90** [comment]
what version of PBSPro are you using?

I believe pbs_mpirun is an mpirun wrapper for MPICH, so it probably won't help with Intel MPI

**Comment by louisXW** [comment]
I tried to figure out which version PBSPro is installed by qstat --version gives: pbs_version = 18.2.3.20181206140456. is this the right way to check?

**Comment by agrawalravi90** [comment]
yes it is, thanks. I think you are logged into a client node (instead of a MoM node) that's why you aren't seeing pbs_tmrsh on your system. Can you try logging into a MoM node and check its pbs_environment?

**Comment by louisXW** [comment]
sorry, what is MoM node? How can I log into a MoM node? I am remote to a server and have no root.

**Comment by agrawalravi90** [comment]
Ok, I think you should contact the admin of your cluster

**Comment by louisXW** [comment]
Here is what I got from the admin of the cluster.
> The I_MPI_HYDRA_BOOTSTRAP and I_MPI_HYDRA_BOOTSTRAP_EXEC environment variables are set in the Intel MPI modules:
> ```
> $ module show composerxe 2>&1 | grep I_MPI_HYDRA_BOOTSTRAP
> setenv           I_MPI_HYDRA_BOOTSTRAP ssh
> setenv           I_MPI_HYDRA_BOOTSTRAP_EXEC pbs_remsh.impi.hydra
> ```
> The pbs_remsh.impi.hydra is a slightly modified (to workaround a bug in parsing ssh optoins) version of pbs_remsh which ships with PBS. This attaches the MPI processes on remote nodes to the PBS MoM daemon.

Does this help?

**Comment by agrawalravi90** [comment]
Thanks for the info. I'm not an expert in this area, but the documentation says that I_MPI_HYDRA_BOOTSTRAP should be set to rsh, not ssh. Also, it might be best to provide the full path to pbs_remsh.impi.hydra in I_MPI_HYDRA_BOOTSTRAP_EXEC. That's all I can think of. If it doesn't work after setting these variables correctly, then I request other members of the community to provide insights on this.
